x/data/server/query_server.go

```diff
@@ -10,55 +15,106 @@ import (
-	//	signers = append(signers, signer)
-	//	iterator.Next()
-	//}
-	//
-	//content := store.Get(DataKey(cid))
-	//
-	//return &data.QueryByCidResponse{
-	//	Timestamp: &timestamp,
-	//	Signers:   signers,
-	//	Content:   content,
-	//}, err
+	iri, err := request.Hash.ToIRI()
+	if err != nil {
+		return nil, err
+	}
+
+	store := ctx.KVStore(s.storeKey)
+	id := s.iriIDTable.GetOrCreateID(store, []byte(iri))
+
+	entry, err := s.getEntry(store, id)
+	if err != nil {
+		return nil, err
+	}
+
+	return &data.QueryByHashResponse{
+		Entry: entry,
+	}, nil
+}
+
+func (s serverImpl) getEntry(store sdk.KVStore, id []byte) (*data.ContentEntry, error) {
+	bz := store.Get(AnchorTimestampKey(id))
+	if len(bz) == 0 {
+		return nil, status.Error(codes.NotFound, "not found")
+	}
+
+	var timestamp gogotypes.Timestamp
+	err := timestamp.Unmarshal(bz)
+	if err != nil {
+		return nil, err
+	}
+
+	var signerEntries []*data.SignerEntry
+	idSignerIndexPrefix := IDSignerIndexPrefix(id)
+	prefixStore := prefix.NewStore(store, idSignerIndexPrefix)
+	iterator := prefixStore.Iterator(nil, nil)
+
+	for iterator.Valid() {
+		signer := sdk.AccAddress(iterator.Key())
+		var timestamp gogotypes.Timestamp
+		err = timestamp.Unmarshal(iterator.Value())
+		if err != nil {
+			return nil, err
+		}
+
+		signerEntries = append(signerEntries, &data.SignerEntry{
+			Signer:    signer.String(),
+			Timestamp: &timestamp,
+		})
+
+		iterator.Next()
+	}
+
+	iri := string(s.iriIDTable.GetValue(store, id))
+	contentHash, _, err := data.ParseIRI(sdk.GetConfig().GetBech32AccountAddrPrefix(), iri)
+	if err != nil {
+		return nil, err
+	}
+
+	entry := &data.ContentEntry{
+		Timestamp: &timestamp,
+		Signers:   signerEntries,
+		Iri:       iri,
+		Hash:      contentHash,
+	}
+
+	rawData := store.Get(RawDataKey(id))
+	if rawData != nil {
+		entry.Content = &data.Content{Sum: &data.Content_RawData{
+			RawData: rawData,
+		}}
+	}
+
+	return entry, nil
 }
 
 func (s serverImpl) BySigner(ctx types.Context, request *data.QueryBySignerRequest) (*data.QueryBySignerResponse, error) {
-	return nil, fmt.Errorf("not implemented")
-	//store := prefix.NewStore(ctx.KVStore(s.storeKey), SignerCIDIndexPrefix(request.Signer))
-	//
-	//var cids [][]byte
-	//pageRes, err := query.Paginate(store, request.Pagination, func(key []byte, value []byte) error {
-	//	cids = append(cids, key)
-	//	return nil
-	//})
-	//if err != nil {
-	//	return nil, err
-	//}
-	//
-	//return &data.QueryBySignerResponse{
-	//	Cids:       cids,
-	//	Pagination: pageRes,
-	//}, nil
+	store := ctx.KVStore(s.storeKey)
+
+	addr, err := sdk.AccAddressFromBech32(request.Signer)
+	if err != nil {
+		return nil, err
+	}
+
+	signerIDStore := prefix.NewStore(store, SignerIDIndexPrefix(addr))
+
+	var entries []*data.ContentEntry
+	pageRes, err := query.Paginate(signerIDStore, request.Pagination, func(key []byte, value []byte) error {
+		entry, err := s.getEntry(store, key)
+		if err != nil {
+			return err
+		}
+
+		entries = append(entries, entry)
```

Comment:
So this looks like this query response will return a paginated list of entries, but each entry could have an unbounded number of signers inside the ContentEntry payload?  Maybe that's fine... but just flagging incase you see that as a potential problem down the line. Not sure what alternatives we have... maybe the `BySigner()` call shouldn't be expected to return ContentEntry's with full lists of all signers inside?

Just a thought.. ignore if you see this as a non-issue.